internal/engine/evaluator.go

```diff
@@ -114,278 +118,264 @@ func (noopEvaluator) Evaluate(_ context.Context, _ tracer.Context, _ *enginev1.C
-				mergedActions.Merge(k.Actions)
-			}
+	pctx := tctx.StartPolicy(fqn)
 
-			if _, ok := activeRoles[r]; !ok {
-				activeRoles[r] = struct{}{}
-				assumedRoles = append(assumedRoles, r)
-			}
-			// The role policy implicitly assumes all parent roles
-			for _, pr := range p.ParentRoles {
-				if _, ok := activeRoles[pr]; !ok {
-					activeRoles[pr] = struct{}{}
-					assumedRoles = append(assumedRoles, pr)
-				}
-			}
-		}
+	// validate the input
+	vr, err := rte.schemaMgr.ValidateCheckInput(ctx, rte.GetSchema(fqn), input)
+	if err != nil {
+		pctx.Failed(err, "Error during validation")
 
-		trail := newAuditTrail(sourceAttrs)
-		result := newEvalResult(input.Actions, trail)
+		return nil, fmt.Errorf("failed to validate input: %w", err)
+	}
+
+	if len(vr.Errors) > 0 {
+		result.ValidationErrors = vr.Errors.SchemaErrors()
 
-		result.AssumedRoles = assumedRoles
+		pctx.Failed(vr.Errors, "Validation errors")
 
-		rpctx := tctx.StartRolePolicyScope(input.Resource.Scope)
+		if vr.Reject {
+			for _, action := range input.Actions {
+				actx := pctx.StartAction(action)
 
-		actions := util.NewGlobMap(mergedActions)
-		for _, a := range input.Actions {
-			actx := rpctx.StartAction(a)
+				result.setEffect(action, EffectInfo{Effect: effectv1.Effect_EFFECT_DENY, Policy: policyKey})
 
-			mappingExists := actions.Get(a) != nil
-			if mappingExists && scopePermission == policyv1.ScopePermissions_SCOPE_PERMISSIONS_OVERRIDE_PARENT {
-				result.setEffect(a, EffectInfo{Effect: effectv1.Effect_EFFECT_ALLOW, Scope: input.Principal.Scope, ActiveRoles: activeRoles})
-				actx.AppliedEffect(effectv1.Effect_EFFECT_ALLOW, "")
-			} else if !mappingExists && scopePermission == policyv1.ScopePermissions_SCOPE_PERMISSIONS_REQUIRE_PARENTAL_CONSENT_FOR_ALLOWS {
-				result.setEffect(a, EffectInfo{Effect: effectv1.Effect_EFFECT_DENY, Policy: noMatchScopePermissions, Scope: input.Principal.Scope, ActiveRoles: activeRoles, IsImplicitDeny: true})
-				actx.AppliedEffect(effectv1.Effect_EFFECT_DENY, fmt.Sprintf("Resource action pair not defined within role policy for resource %s and action %s", input.Resource.Kind, a))
+				actx.AppliedEffect(effectv1.Effect_EFFECT_DENY, "Rejected due to validation failures")
 			}
+			return result, nil
 		}
+	}
 
-		result.setDefaultEffect(rpctx, EffectInfo{Effect: effectv1.Effect_EFFECT_NO_MATCH, ActiveRoles: activeRoles})
+	evalCtx := newEvalContext(rte.evalParams, request)
 
-		return result, nil
-	})
-}
+	effectiveDerivedRoles := make(internal.StringSet)
+	// This is a bit of a bolt-on evaluation in order to support the existing requirement of evaluating all derived roles
+	// defined within a resource policy (regardless of if they're used or not). This is primarily for the
+	// runtime.EffectiveDerivedRoles usage
+	includingParentRoles := make(map[string]struct{})
+	for _, r := range rte.GetParentRoles(input.Principal.Roles) {
+		includingParentRoles[r] = struct{}{}
+	}
+	for _, scope := range scopes {
```

Comment:
This puzzled me as well! I originally didn't have this loop in place because it felt _wrong_ to me, but we have a bunch of expectations in tests that require it to be here 🤔

I would expect `effectiveDerivedRoles` to be those which are defined within a rule that's matched and evaluated to produce an EFFECT. However, that doesn't seem to be existing behaviour--it seems that it includes any derived roles that satisfy the runtime constraints irrespective of whether or not the derived role is present in an effective rule (there's an [initial loop run before evaluation](https://github.com/cerbos/cerbos/blob/main/internal/engine/evaluator.go#L252-L286)).

On further reading, I assumed that this was the case because we require all _potentially_ effective derived roles to be made available for the `runtime.EffectiveDerivedRoles` feature?

Reply:
There's a subtle distinction between `runtime.EffectiveDerivedRoles` and `meta.EffectiveDerivedRoles`.

In the current engine, each policy in the policy set is evaluated one at a time. It grabs the policy to evaluate, gets all its derived roles, finds out which ones are activated, and sets that list as `runtime.EffectiveDerivedRoles`. That gets reset for the next policy in the set. Whereas, `meta.EffectiveDerivedRoles` is the set of all effective derived roles we've encountered during the evaluation of the entire request. Again, it's not necessarily _every_ derived role from all the policies because evaluation could stop early (e.g. the leaf policy in a scoped set is enough to make all the decisions and we never needed to go up the hierarchy and evaluate the parents). 

If we are really unlucky, Hyrum's law would dictate that someone is relying on this property of `runtime.EffectiveDerivedRoles` resetting each time. But, the chances are minuscule and I think that it's an acceptable breaking change for us to make. However, I do know that some users rely on `meta.EffectiveDerivedRoles` to make some decisions on the client side. So, changing what that means is probably not quite so acceptable.